src/cbor_decode.c

```diff
@@ -0,0 +1,390 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <string.h>
+#include "cbor_decode.h"
+
+static uint32_t additional_len(uint8_t additional)
+{
+	switch (additional) {
+	case 24: return 1;
+	case 25: return 2;
+	case 26: return 4;
+	case 27: return 8;
+	default: return 0;
+	}
+}
+
+#define COMP_PTRS(type, p_res, p_min, p_max) \
+		(((p_min == NULL) || (*(type *)p_res >= *(type *)p_min)) \
+		&& ((p_max == NULL) || (*(type *)p_res <= *(type *)p_max)))
+
+static bool value_extract(uint8_t const **const pp_payload,
+		uint8_t const *const p_payload_end, size_t *const p_elem_count,
+		void * const p_result, size_t result_len)
+{
+	uint8_t *p_u8_result  = (uint8_t *)p_result;
+	uint8_t additional = **pp_payload & 0x1F;
+	(*pp_payload)++;
+
+	cbor_decode_trace();
+
+	if (result_len == 0 || *p_elem_count == 0) {
```

Comment:
Own static function for "rollback and return error". the content inside the if-sentence is duplicated many places

Reply:
Added macros